File: Assignment-6/1/MinHeap.java
import java.util.Scanner;
import java.util.ArrayList;
import java.lang.Long;
import java.lang.Integer;

public class MinHeap
{
    private static int leftChild(int i)
    {
        return (2 * i) + 1;
    }

    private static int rightChild(int i)
    {
        return (2 * i) + 2;
    }

    private static void swap(int a, int b, ArrayList<Long> mHeap)
    {
        Long temp = mHeap.get(a);
        mHeap.set(a, mHeap.get(b));
        mHeap.set(b, temp);
    }

    private static void swiftDown(int i, int SIZE,ArrayList<Long> mHeap, ArrayList<Integer> mSwaps)
    {
        int leftChildIndex = leftChild(i);
        int rightChildIndex = rightChild(i);
        int minIndex = i;

        if(leftChildIndex < SIZE && mHeap.get(minIndex) > mHeap.get(leftChildIndex))
            minIndex = leftChildIndex;
        
        if(rightChildIndex < SIZE && mHeap.get(minIndex) > mHeap.get(rightChildIndex))
            minIndex = rightChildIndex;
        
        if(i != minIndex)
        {
            mSwaps.add(i);
            mSwaps.add(minIndex);
            swap(minIndex, i, mHeap);
            swiftDown(minIndex, SIZE, mHeap, mSwaps);
        }
    }

    public static void main(String[] args) 
    {
        int SIZE = 0;
        ArrayList<Long> mHeap = new ArrayList<Long>();
        ArrayList<Integer> mSwaps = new ArrayList<Integer>();
        Scanner mScanner = new Scanner(System.in);
        SIZE = mScanner.nextInt();

        for(int i = 0; i < SIZE; i++)
        {
            mHeap.add(mScanner.nextLong());
        }

        for(int i = (SIZE/2); i >= 0; i--)
        {
            swiftDown(i, SIZE, mHeap, mSwaps);
        }
        
        System.out.println((mSwaps.size()/2));
        for(int i = 0; i < mSwaps.size(); i++)
        {
            System.out.println(mSwaps.get(i++) + " " + mSwaps.get(i));
        }
        mScanner.close();
    }
}

File: Assignment-6/1/min_heap.cpp
#include<iostream>
#include<vector>

/*
  ************BINARY MIN HEAP*************
  * It is a refactored method for selection sort.
  * Definition :
    Binary min-heap is a binary tree (each node has zero , one or more children) where the value of
    each node is atmost the values of its children.
  * Implementation of Binary min-heap in Arrays is done here.
*/

using namespace std;

// Global variables

long long SIZE = 0;
vector<long long> min_heap;
vector<long long> swaps;

// This will swap two numbers in the array.
void swap(long long &a, long long &b)
{
  long long temp = a;
  a = b;
  b = temp;
}

// This will return the index of parent of node i.
long long parent(long long i)
{
  return (i/2);
}

// This will return the index of left child of node i.
long long left_child(long long i)
{
  return (2 * i) + 1;
}

// This will return the index of right child of node i.
long long right_child(long long i)
{
  return (2 * i) + 2;
}

/*
 This will move a node down in the tree if it does not satisfy the
 binary min-heap property.

 -> Checks whether the left and right child of the node is lesser than the node i
 -> Swaps it
 -> Perform swift_down() operation on the swapped node.
*/
void swift_down(long long i)
{
  long long min_index = i;
  long long left_child_index = left_child(i);
  long long right_child_index = right_child(i);

  if(left_child_index < SIZE && min_heap[left_child_index] < min_heap[min_index])
    min_index = left_child_index;

  if(right_child_index < SIZE && min_heap[right_child_index] < min_heap[min_index])
    min_index = right_child_index;

  if(i != min_index)
  {
    swaps.push_back(i);
    swaps.push_back(min_index);
    swap(min_heap[min_index], min_heap[i]);
    swift_down(min_index);
  }
}

int main()
{
  cin>>SIZE;
  long long temp = 0;

  // Getting array items.
  for(long long i = 0; i < SIZE; i++)
  {
    cin>>temp;
    min_heap.push_back(temp);
  }

  // swift_down() for n/2 to 0 nodes.
  for(long long i = (SIZE / 2); i >= 0; i--)
  {
    swift_down(i);
  }

  cout<<(swaps.size() / 2)<<endl;
  for(long long i = 0; i < swaps.size(); i++)
    cout<<swaps[i]<<" "<<swaps[++i]<<endl;
  //cout<<endl;

  /*
    // Implementation of Heap sort (Ascending order) is done here.

    for(long long i = 0; i < SIZE; i++)
    {
      cout<<min_heap[i]<<" ";
    }
    cout<<endl;
  */
  return 0;
}
